### ISSUE TITLE
PrivateBin "processes/eats"HTML,  PHP, CSS Code

### ISSUE THREAD

**Issue by xshadow** [PrivateBin/PrivateBin#403]
## Steps to reproduce
1. Go to https://privatebin.net 
2. Enter : 
```
<?php
echo "Hello PrivateBin!";
?>
```
3. Send it
4. Click on the provided link
5. Click on "Raw Text"
6. See white page

Example: https://privatebin.net/?39bbb18ab72aa5e8#3B2bk8/c/VCU3KBn0JrpNB9jy1aOE62NjjQ/hTGNiEc=

### What happens

It seems that PrivateBin somehow processes the PHP code and displays a blank webpage. It's a little unclear what happens in the background. 

Source code output of blank webpage: 
```
<!DOCTYPE html><html><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1"><meta name="robots" content="noindex"><meta name="referrer" content="no-referrer"><title>PrivateBin</title><link rel="apple-touch-icon" href="img/apple-touch-icon.png?1.2.1" sizes="180x180"><link rel="icon" type="image/png" href="img/favicon-32x32.png?1.2.1" sizes="32x32"><link rel="icon" type="image/png" href="img/favicon-16x16.png?1.2.1" sizes="16x16"><link rel="manifest" href="manifest.json?1.2.1"><link rel="mask-icon" href="img/safari-pinned-tab.svg?1.2.1" color="#ffcc00"><link rel="shortcut icon" href="img/favicon.ico"><meta name="msapplication-config" content="browserconfig.xml"><meta name="theme-color" content="#ffe57e"></head><body><pre></pre></body></html>
```

### What should happen

The pasted PHP code should be shown as raw text. 

## Basic information

**Webserver:** https://privatebin.net/

**Browser:** Tested in Firefox 64.0 and Chromium Version 71.0.3578.98 

**PrivateBin version:** 1.2.1

I can reproduce this issue on <https://privatebin.net>: Yes

**Comment by r4sas** [comment]
Hah, seems like raw text is not preprocessed with htmlspecialchars (PHP) command.

**Comment by elrido** [comment]
Of course this has nothing to do with PHP, as the paste is encrypted with a key not know to the server. The decryption and display is fulyl handled on the JS side in the browser. In this particular case I assume that the DOMpurify library detects the use of the PHP tags as malicious code and filters them out when switching to the raw text mode, leaving an empty text to be displayed.

The real question is, what we do differently when displaying the document in the UI vs the raw text mode (which uses an HTML document with a pre tag to avoid issues when manipulating the browser history, so the back button still works).

**Comment by alxpettit** [comment]
I could not replicate this issue. Using Chromium 71.0.3578.98.

**Comment by xshadow** [comment]
It's not only PHP it also renders other code.

[Look here](https://privatebin.net/?73fb530856fac7cd#GkyFNEFTLlYzD1/woIkbvuiayXWNP2CH7U+88hmesPo=) - and click "Raw text"

**Comment by r4sas** [comment]
Maybe we need escape special chars before sending them to DOMpurify?
https://github.com/PrivateBin/PrivateBin/blob/48560e3d606032038e085d7e8e159657fa6864ce/js/privatebin.js#L2928

https://stackoverflow.com/a/4835406

**Comment by gwern** [comment]
It also eats HTML/CSS tags which are in quotes too. Example: half of https://privatebin.net/?3c76f4e4ceaf3dc2#LnEnYGSE7GKHEsB58IUIlEqzq2LJ2orrjHVtyfcOwqk= is deleted in 'raw'.

**Comment by r4sas** [comment]
It eats everything what looks like HTML tags.

**Comment by rugk** [comment]
> Maybe we need escape special chars before sending them to DOMpurify?

Thinking about it, yes, that may be what we want…

The thing is DOMPurify allows HTML (as HTML) that is not a security problem. However, in any "raw text" mode, we do not want to show it as HTML, but as raw text…
The thing why it works in the preview is likely only because of the `pre` tag.

The only case, where we do not want to replace all special chars is when we show a Markdown preview. There, safe HTML is allowed.

**Comment by Zitronella** [comment]
i wrote in paste `https://example.org/afu.php`
https://example.org/afu.php
(i see also github eats the link so i put it here again in code)
Paste is empty. Y
ou only can see the link in raw text
https://privatebin.net/?59b70cbb26cc06dc#isCxPKvHV1qxY75NiJv+Ua0U2LCCyvkn6S9GfdzquGY=
![github-eats-afu-link-write](https://user-images.githubusercontent.com/14326582/55557042-45cde180-56e9-11e9-98f8-54d7b2dbb5e0.png)
![github-eats-afu-link-preview](https://user-images.githubusercontent.com/14326582/55557112-685ffa80-56e9-11e9-9ea9-100a7072998c.png)

Maybe it helps to find out and fix it.

**Comment by doobry-systemli** [comment]
Any news? It would be awesome to get this fixed :)

**Comment by rugk** [comment]
> Maybe we need escape special chars before sending them to DOMpurify?

Yes, I guess so. We had this in the past and yeah… we should do it _before_ sending to DOMPurify of course.

**Comment by elrido** [comment]
Should maybe be reviewed before releasing 1.3?

**Comment by elrido** [comment]
Sorry, I forgot: This was already brought up in #358 and got resolved in 8b71cb0.
